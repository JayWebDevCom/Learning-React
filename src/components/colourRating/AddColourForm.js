import React from 'react';
import {colorActions} from "../../actions/colorActions";

export const AddColourForm = ({store}) => {
    let _title, _colour;

    const submit = e => {
        e.preventDefault();
        store.dispatch(colorActions.addColor(_title.value, _colour.value));
        _title.value = '';
        _colour.value = '#000000';
        _title.focus()
    };

    return (
        <form onSubmit={submit} id={'add-color-form'}>
            <input ref={input => _title = input}
                   type='text'
                   placeholder="colour title..." required/>
            <input ref={input => _colour = input}
                   type='color' required/>
            <button>Add</button>
        </form>
    )
};