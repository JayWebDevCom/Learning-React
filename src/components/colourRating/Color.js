import React, {Component} from 'react';
import {StarRating} from "./StarRating";
import PropTypes from "prop-types";

export class Color extends Component {

    UNSAFE_componentWillMount() {
        this.style = {
            backgroundColor: '#CCC'
        }
    }

    UNSAFE_componentWillUpdate(nextProps, nextState, nextContext) {
        this.style = null
    }

    shouldComponentUpdate(nextProps, nextState, nextContext) {
        const {rating} = this.props;
        return rating !== nextProps.rating
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        const {title, rating} = this.props;
        const status = (rating > prevProps.rating) ? 'better' : 'worse' ;
        console.log(`${title} is getting a ${status}`)
    }

    render() {
        const {title, color, rating = 0, onRemove = f => f, onRate = f => f} = this.props;
        return (
            <section className={'color'} style={this.style}>
                <h1 ref={"title"}>{title}</h1>
                <button onClick={onRemove}>X</button>
                <div className={'color'}
                     style={{backgroundColor: color}}
                >
                </div>
                <div>
                    <StarRating starsSelected={rating} onRate={onRate}/>
                </div>
            </section>
        )
    }
}

Color.propTypes = {
    title: PropTypes.string.isRequired,
    color: PropTypes.string.isRequired,
    rating: PropTypes.number,
    onRemove: PropTypes.func,
    onRate: PropTypes.func
};

Color.defaultProps = {
    title: undefined,
    rating: 0,
    color: "#000000",
    onRate: f => f
};