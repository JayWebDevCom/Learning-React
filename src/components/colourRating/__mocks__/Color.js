import React from 'react';

export const Color = ({rating, onRate = f => f, onRemove = f => f}) =>
    <div className={'mock-color'}>
        <button className={'rate'} onClick={() => onRate(rating)}/>
        <button className={'remove'} onClick={() => onRemove(rating)}/>
    </div>;
