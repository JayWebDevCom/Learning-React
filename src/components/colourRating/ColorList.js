import React from 'react'
import {Color} from './Color'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'
import {colorActions} from "../../actions/colorActions";

const InternalColorList = ({store}) => {

    const {colors} = store.getState();

    return (
        <div className='color-list'>
            {(colors.length === 0) ?
                <p>No Colours Listed. (Add a Color)</p> :
                colors.map(color =>
                    <Color key={color.id}
                           {...color}
                           onRate={(rating) => store.dispatch(colorActions.rateColor(color.id, rating))}
                           onRemove={() => store.dispatch(colorActions.removeColor(color.id))}
                    />
                )
            }
        </div>
    );
};

const mapStateToProps = state => ({
        colors: state.colors
    });

export const ColorList = connect(mapStateToProps)(InternalColorList);

ColorList.propTypes = {
    colors: PropTypes.array,
    onRate: PropTypes.func,
    onRemove: PropTypes.func
};