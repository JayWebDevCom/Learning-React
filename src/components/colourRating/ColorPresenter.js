import React from 'react';
import colorsFile from "./colorsFile"
import {v4} from "uuid";
import {AddColourForm} from "./AddColourForm";
import {ColorList} from "./ColorList";

export class ColorPresenter extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            colors: colorsFile.colors
        };
        this.addColor = this.addColor.bind(this);
        this.rateColor = this.rateColor.bind(this);
        this.removeColor = this.removeColor.bind(this);
    }

    addColor = (title, color) => {
        this.setState(prevState => ({
            colors: [
                ...prevState.colors,
                {
                    id: v4(),
                    title,
                    color,
                    rating: 0
                }
            ]
        }))
    };

    rateColor = (id, rating) => {
        this.setState(prevState => ({
            colors: prevState.colors.map(color =>
                (color.id !== id) ?
                    color :
                    {
                        ...color,
                        rating
                    }
            )
        }));
    };

    removeColor = (id) => {
        this.setState(prevState => ({
            colors: prevState.colors.filter(
                color => color.id !== id
            )
        }))
    };

    render() {
        const {addColor, rateColor, removeColor} = this;
        const {colors} = this.state;

        return (
            <div>
                <AddColourForm onNewColour={addColor}/>
                <ColorList colors={colors} onRate={rateColor} onRemove={removeColor}/>
            </div>
        )
    }
}
