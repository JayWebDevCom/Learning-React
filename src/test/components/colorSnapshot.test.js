import React from 'react';
import {shallow} from 'enzyme';
import {Color} from '../../components/colourRating/Color'
import toJson from "enzyme-to-json";
import {compose} from "redux";

describe('<Color /> Snapshot', () => {

    it('renders correct props', () => {
        const shallowExpect =
            compose(expect, toJson, shallow);

        shallowExpect(<Color title={'color test'} color={'#R4R4R4'} rating={3}/>)
            .toMatchSnapshot();

    });
});