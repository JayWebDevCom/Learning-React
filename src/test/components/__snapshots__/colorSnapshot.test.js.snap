// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Color /> Snapshot renders correct props 1`] = `
<section
  className="color"
  style={
    Object {
      "backgroundColor": "#CCC",
    }
  }
>
  <h1>
    color test
  </h1>
  <button
    className="remove"
    onClick={[Function]}
  >
    X
  </button>
  <div
    className="color"
    style={
      Object {
        "backgroundColor": "#R4R4R4",
      }
    }
  />
  <div>
    <StarRating
      onRate={[Function]}
      starsSelected={3}
    />
  </div>
</section>
`;
